presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -228,14 +237,34 @@ public IPrestoSparkTaskExecutor create(
         PrestoSparkRowBuffer rowBuffer = new PrestoSparkRowBuffer(memoryManager);
 
         ImmutableMap.Builder<PlanNodeId, Iterator<PrestoSparkRow>> shuffleInputs = ImmutableMap.builder();
+        ImmutableMap.Builder<PlanNodeId, Iterator<PrestoSparkSerializedPage>> broadcastInputs = ImmutableMap.builder();
         for (RemoteSourceNode remoteSource : fragment.getRemoteSourceNodes()) {
-            ImmutableList.Builder<Iterator<PrestoSparkRow>> remoteSourceInputs = ImmutableList.builder();
+            List<Iterator<PrestoSparkRow>> shuffleRemoteSourceInputs = new ArrayList<>();
+            List<Iterator<PrestoSparkSerializedPage>> broadcastRemoteSourceInputs = new ArrayList<>();
             for (PlanFragmentId sourceFragmentId : remoteSource.getSourceFragmentIds()) {
-                Iterator<Tuple2<Integer, PrestoSparkRow>> input = inputs.getShuffleInputs().get(sourceFragmentId.toString());
-                checkArgument(input != null, "input is missing for fragmentId: %s", sourceFragmentId);
-                remoteSourceInputs.add(Iterators.transform(input, tuple -> tuple._2));
+                Iterator<Tuple2<Integer, PrestoSparkRow>> shuffleInput = inputs.getShuffleInputs().get(sourceFragmentId.toString());
+                if (shuffleInput != null) {
+                    shuffleRemoteSourceInputs.add(Iterators.transform(shuffleInput, tuple -> tuple._2));
+                    continue;
+                }
+                Broadcast<List<PrestoSparkSerializedPage>> broadcastInput = inputs.getBroadcastInputs().get(sourceFragmentId.toString());
+                if (broadcastInput != null) {
+                    // TODO: Enable NullifyingIterator once migrated to one task per JVM model
+                    // NullifyingIterator removes element from the list upon return
+                    // This allows GC to gradually reclaim memory
+                    // broadcastRemoteSourceInputs.add(getNullifyingIterator(broadcastInput.value()));
+                    broadcastRemoteSourceInputs.add(broadcastInput.value().iterator());
+                    continue;
+                }
+                throw new IllegalArgumentException("Input not found for sourceFragmentId: " + sourceFragmentId);
+            }
+            verify(shuffleRemoteSourceInputs.isEmpty() || broadcastRemoteSourceInputs.isEmpty(), "single remote source cannot accept both, broadcast and shuffle inputs");
```

Comment:
Because of the `continue` in line 248, `shuffleRemoteSourceInputs` and `broadcastRemoteSourceInputs` cannot both be non-empty right? -- Maybe check `shuffleInputs` and `broadcastInput` cannot both present earlier?

Reply:
Makes sense. Let me do that.